patches/api/0447-Add-FluidState-API.patch

```diff
@@ -0,0 +1,171 @@
++     * Returns whether this fluid is lava.
++     *
++     * @return true if the fluid is lava, false otherwise
++     */
++    boolean isLava();
++}
+diff --git a/src/main/java/io/papermc/paper/block/fluid/type/FallingFluid.java b/src/main/java/io/papermc/paper/block/fluid/type/FallingFluid.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..360a27fca76b0e63aa458aa3425a987224a4ac79
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/block/fluid/type/FallingFluid.java
+@@ -0,0 +1,13 @@
++package io.papermc.paper.block.fluid.type;
++
++import io.papermc.paper.block.fluid.FluidData;
++
++public interface FallingFluid extends FluidData {
++
++    /**
++     * Get if this liquid is falling.
++     *
++     * @return true if falling
++     */
++    boolean isFalling();
++}
+diff --git a/src/main/java/io/papermc/paper/block/fluid/type/Flowing.java b/src/main/java/io/papermc/paper/block/fluid/type/Flowing.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..14589b2de95e0f679fe9785e06648ace56449885
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/block/fluid/type/Flowing.java
+@@ -0,0 +1,26 @@
++package io.papermc.paper.block.fluid.type;
++
++public interface Flowing extends FallingFluid {
++
++    /**
++     * Get the level of the flowing liquid.
++     *
++     * @return the level
++     */
++    int getLevel();
++
++    /**
++     * Get the minimum possible level for the liquid.
++     *
++     * @return the minimum level
++     */
++    int getMinimumLevel();
++
++    /**
++     * Get the maximum possible level for the liquid.
++     *
++     * @return the maximum level
++     */
++    int getMaximumLevel();
++
++}
+diff --git a/src/main/java/org/bukkit/RegionAccessor.java b/src/main/java/org/bukkit/RegionAccessor.java
+index cbb51dde78f792db4ddac43f144d23ff4b12d25f..440af5ed52bed8fac0d9fa47b9cb1eaf1716f8bd 100644
+--- a/src/main/java/org/bukkit/RegionAccessor.java
++++ b/src/main/java/org/bukkit/RegionAccessor.java
+@@ -102,6 +102,28 @@ public interface RegionAccessor extends Keyed { // Paper
+      */
+     @NotNull
+     BlockState getBlockState(int x, int y, int z);
++    // Paper Start - FluidState API
++
++    /**
++     * Gets the {@link io.papermc.paper.block.fluid.FluidData} at the specified position.
++     *
++     * @param x The x-coordinate of the position
++     * @param y The y-coordinate of the position
++     * @param z The z-coordinate of the position
++     * @return The {@link io.papermc.paper.block.fluid.FluidData} at the specified position
++     */
++    @NotNull
++    public io.papermc.paper.block.fluid.FluidData getFluidAt(int x, int y, int z);
++
++    /**
++     * Gets the {@link io.papermc.paper.block.fluid.FluidData} at the given {@link Location}
++     *
++     * @param location The location of the fluid
++     * @return The fluid data at the given location
++     */
++    @NotNull
++    public io.papermc.paper.block.fluid.FluidData getFluidAt(Location location);
```

Comment:
Maybe replace both of these methods with a single one taking a [Position](https://jd.papermc.io/paper/1.20/io/papermc/paper/math/Position.html)?

You then can also use `MCUtil.toBlockPos` to convert the position into a NMS BlockPosition

Reply:
Why not use Location just like all other methods like getBlockAt?

Reply:
Position is a new API that is used where a world from the location object does not get used, as methods on the World / RegionAccessor discard the world inside the location because it uses itself

Reply:
Fair enough. I sill think a method that takes x, y, z is always nice to have.